### PR TITLE
Add expected seal duration to custom deal logic

### LINKED ISSUE #2048: storage miner should reject deals that it can't seal in time
If a miner knows it can't seal the deal into a sector by the deal start time, it should reject the deal

### PR DESCRIPTION
**Pull request by arajasek**
And reject deals that start too early.

Fixes #2048

### CODE REVIEW

File: api/api_storage.go [hunk 1]
```diff
@@ -46,6 +46,10 @@ type StorageMiner interface {
 	// SectorGetSealDelay gets the time that a newly-created sector
 	// waits for more deals before it starts sealing
 	SectorGetSealDelay(context.Context) (time.Duration, error)
+	// SectorSetExpectedSealDuration sets the expected time for a sector to seal
+	SectorSetExpectedSealDuration(context.Context, time.Duration) error
```
Comment (magik6k) [inline api/api_storage.go:50]: We should probably create a new `Sealing` namespace for some of those functions, `Sector` is a bit weird here

Comment (arajasek) [reply api/api_storage.go:50]: Yeah, strong agree...you okay with that being a separate PR?

Comment (magik6k) [reply api/api_storage.go:50]: Yep

File: node/modules/storageminer.go [hunk 2]
```diff
@@ -335,14 +335,26 @@ func NewStorageAsk(ctx helpers.MetricsCtx, fapi lapi.FullNode, ds dtypes.Metadat
...
+	blocklistFunc dtypes.StorageDealPieceCidBlocklistConfigFunc,
+	expectedSealTimeFunc dtypes.GetExpectedSealDurationFunc) (storagemarket.StorageProvider, error) {
 	net := smnet.NewFromLibp2pHost(h)
 	store, err := piecefilestore.NewLocalFileStore(piecefilestore.OsPath(r.Path()))
 	if err != nil {
 		return nil, err
 	}
 
-	opt := storageimpl.CustomDealDecisionLogic(func(ctx context.Context, deal storagemarket.MinerDeal) (bool, string, error) {
+	opt := storageimpl.CustomDealDecisionLogic(func(ctx context.Context, deal storagemarket.MinerDeal, ht abi.ChainEpoch) (bool, string, error) {
```
Comment (hannahhoward) [inline node/modules/storageminer.go:357]: I haven't looked closely at this function but it looks like it's getting pretty long to live in a constructor. I feel like you might just want to create a struct (which could store the current SealDuration in it it) and have one of its methods be the deal decider?

Also, are you sure you don't want us to expose the config options on StorageProvider after construction, so they can be changed later?

**Comment by hannahhoward** [comment]
my review is not blocking but I'd like to not add the height to deal decider unless it's absolutely neccesary.

File: node/modules/storageminer.go [hunk 3]
```diff
@@ -335,7 +335,19 @@ func NewStorageAsk(ctx helpers.MetricsCtx, fapi lapi.FullNode, ds dtypes.Metadat
...
+	h host.Host, ds dtypes.MetadataDS,
+	ibs dtypes.StagingBlockstore,
+	r repo.LockedRepo,
+	pieceStore dtypes.ProviderPieceStore,
+	dataTransfer dtypes.ProviderDataTransfer,
+	spn storagemarket.StorageProviderNode,
+	onlineOk dtypes.ConsiderOnlineStorageDealsConfigFunc,
+	offlineOk dtypes.ConsiderOfflineStorageDealsConfigFunc,
+	blocklistFunc dtypes.StorageDealPieceCidBlocklistConfigFunc,
+	expectedSealTimeFunc dtypes.GetExpectedSealDurationFunc) (storagemarket.StorageProvider, error) {
```
Comment (magik6k) [inline node/modules/storageminer.go:350]: Probably warrants a struct with `fx.In` (not blocking this PR)

**Pull request merged**